### PR TITLE
Expose certificate for HTTPS responses

### LINKED ISSUE #2726: Expose certificate information on HTTPS responses
For scraper I'm working on, I'd like to have access to the certificate supplied by the server for HTTPS requests. As far as I can tell, this is not currently exposed on either requests or responses.

### PR DESCRIPTION
**Pull request by elacuesta**
Fixes #2726

The approach I took here is very similar to the one in #3940.

~~No tests at the moment (not sure how to test this without requesting external pages since the mockserver uses HTTP, any help is appreciated :heart:),~~ posting a `scrapy shell` session instead:
```
$ scrapy shell https://example.org
2019-10-01 14:23:58 [scrapy.utils.log] INFO: Scrapy 1.7.0 started (bot: scrapybot)
(...)
Available Scrapy objects:
[s]   scrapy     scrapy module (contains scrapy.Request, scrapy.Selector, etc)
[s]   crawler    <scrapy.crawler.Crawler object at 0x7fdfc53b0208>
[s]   item       {}
[s]   request    <GET https://example.org>
[s]   response   <200 https://example.org>
[s]   settings   <scrapy.settings.Settings object at 0x7fdfc1147898>
[s]   spider     <DefaultSpider 'default' at 0x7fdfc073af98>
[s] Useful shortcuts:
[s]   fetch(url[, redirect=True]) Fetch URL and update local objects (by default, redirects are followed)
[s]   fetch(req)                  Fetch a scrapy.Request and update local objects 
[s]   shelp()           Shell help (print this help)
[s]   view(response)    View response in a browser
In [1]: response.certificate                                                                                                                                                                                                                  
Out[1]: <Certificate Subject=b'www.example.org' Issuer=b'DigiCert SHA2 Secure Server CA'>

In [2]: response.certificate.getIssuer()                                                                                                                                                                                                      
Out[2]: <DN 'commonName': b'DigiCert SHA2 Secure Server CA', 'organizationName': b'DigiCert Inc', 'countryName': b'US'>

In [3]: print(response.certificate.getIssuer().inspect())                                                                                                                                                                                     
               Common Name: DigiCert SHA2 Secure Server CA
              Country Name: US
         Organization Name: DigiCert Inc
```

### CODE REVIEW

File: docs/topics/request-response.rst [hunk 1]
```diff
@@ -581,7 +581,7 @@ Sending a JSON POST request with a JSON payload::
 Response objects
 ================
 
-.. class:: Response(url, [status=200, headers=None, body=b'', flags=None, request=None])
+.. autoclass:: Response
```
Comment (kmike) [inline docs/topics/request-response.rst]: could you please separate Response doc changes to a different PR, to allow merging them separately?

**Comment by kmike** [comment]
We can have https mockserver as well if we want, e.g. https://github.com/scrapinghub/splash/blob/master/splash/tests/mockserver.py has it (if I'm not mistaken, it was originally based on Scrapy's, and evolved since then)

File: docs/topics/request-response.rst (hunk 1 above)
Comment (elacuesta) [reply docs/topics/request-response.rst]: Done, please see #4055. Thanks!

**Comment by elacuesta** [comment]
@kmike I'm having trouble with the mockserver in HTTPS mode. Sadly, I don't have a lot of experience with self-signed certificates (I assume that is the case).

```diff
diff --git a/tests/test_crawl.py b/tests/test_crawl.py
index 3fc13eeb..94d2d8ec 100644
--- a/tests/test_crawl.py
+++ b/tests/test_crawl.py
@@ -3,6 +3,7 @@ import logging
 
 from testfixtures import LogCapture
 from twisted.internet import defer
+from twisted.internet.ssl import Certificate
 from twisted.trial.unittest import TestCase
 
 from scrapy.http import Request
@@ -277,3 +278,19 @@ with multiples lines
 
         self._assert_retried(log)
         self.assertIn("Got response 200", str(log))
+
+    @defer.inlineCallbacks
+    def test_response_ssl_certificate_none(self):
+        crawler = self.runner.create_crawler(SingleRequestSpider)
+        url = self.mockserver.url("/status?n=200", is_secure=False)
+        yield crawler.crawl(seed=url, mockserver=self.mockserver)
+        self.assertIsNone(crawler.spider.meta['responses'][0].certificate)
+
+    @defer.inlineCallbacks
+    def test_response_ssl_certificate(self):
+        crawler = self.runner.create_crawler(SingleRequestSpider)
+        url = self.mockserver.url("/status?n=200", is_secure=True)
+        yield crawler.crawl(seed=url, mockserver=self.mockserver)
+        cert = crawler.spider.meta['responses'][0].certificate
+        print('Response.certificate:', cert)
+        self.assertIsInstance(cert, Certificate)
```

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.test_crawl.CrawlTestCase testMethod=test_response_ssl_certificate>, msg = "None is not an instance of <class 'twisted.internet._sslverify.Certificate'>"

    def fail(self, msg=None):
        """
        Absolutely fail the test.  Do not pass go, do not collect $200.
    
        @param msg: the message that will be displayed as the reason for the
        failure
        """
>       raise self.failureException(msg)
E       twisted.trial.unittest.FailTest: None is not an instance of <class 'twisted.internet._sslverify.Certificate'>

/.../scrapy/.tox/py35/lib/python3.5/site-packages/twisted/trial/_synctest.py:375: FailTest
------------------------------------------------------------------------------------------------------------ Captured stdout call ------------------------------------------------------------------------------------------------------------
Response.certificate: None
------------------------------------------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------------------------------------------
Coverage.py warning: --include is ignored because --source is set (include-ignored)
------------------------------------------------------------------------------------------------------------- Captured log call --------------------------------------------------------------------------------------------------------------
WARNING  scrapy.core.downloader.tls:tls.py:90 Ignoring error while verifying certificate from host "0.0.0.0" (exception: ValueError('Invalid DNS-ID.',))
============================================================================================================== warnings summary ==============================================================================================================
tests/test_crawl.py::CrawlTestCase::test_crawl_multiple
  /.../scrapy/scrapy/core/downloader/webclient.py:4: DeprecationWarning: twisted.web.client.HTTPClientFactory was deprecated in Twisted 16.7.0: please use https://pypi.org/project/treq/ or twisted.web.client.Agent instead
    from twisted.web.client import HTTPClientFactory

tests/test_crawl.py::CrawlTestCase::test_response_ssl_certificate
  /.../scrapy/scrapy/core/downloader/contextfactory.py:53: DeprecationWarning: Passing method to twisted.internet.ssl.CertificateOptions was deprecated in Twisted 17.1.0. Please use a combination of insecurelyLowerMinimumTo, raiseMinimumTo, and lowerMaximumSecurityTo instead, as Twisted will correctly configure the method.
    acceptableCiphers=self.tls_ciphers)

tests/test_crawl.py::CrawlTestCase::test_response_ssl_certificate
  /.../scrapy/.tox/py35/lib/python3.5/site-packages/service_identity/pyopenssl.py:49: SubjectAltNameWarning: Certificate with CN 'localhost' has no `subjectAltName`, falling back to check for a `commonName` for now.  This feature is being removed by major browsers and deprecated by RFC 2818.  service_identity will remove the support for it in mid-2018.
    cert_patterns=extract_ids(connection.get_peer_certificate()),

-- Docs: https://docs.pytest.org/en/latest/warnings.html
```

**Comment by Gallaecio** [comment]
@elacuesta Could you push your WIP for certificate tests?

**Comment by elacuesta** [comment]
> @elacuesta Could you push your WIP for certificate tests?

@Gallaecio Pushed, sorry I missed this comment

**Comment by Gallaecio** [comment]
The issue is in the Scrapy code, not the mock server. When the response body is empty, the certificate is set to `None` in `_cb_bodyready`.

I tried reading it from `txresponse._transport._producer.getPeerCertificate()` at that point, but I get `AttributeError("'NoneType' object has no attribute 'getPeerCertificate'",)`.

So:
* Add copies of your tests that use the `/echo` endpoint instead of `/status?n=200`.
* Fix the empty body issue if you know how or feel like it. Otherwise, I’m OK with marking that test as an expected failure, and after merging these changes creating a new ticket to eventually fix that scenario.

**Comment by Gallaecio** [comment]
I think we should also keep tests with `/status?n=200`, and either mark the SSL one as an expected failure or fix the issue.

File: scrapy/core/downloader/handlers/http11.py [hunk 2]
```diff
@@ -445,6 +445,14 @@ def __init__(self, finished, txresponse, request, maxsize, warnsize, fail_on_dat
         self._fail_on_dataloss_warned = False
         self._reached_warnsize = False
         self._bytes_received = 0
+        self._certificate = None
+
+    def connectionMade(self):
+        try:
+            assert self._certificate is None
```
Comment (Gallaecio) [inline scrapy/core/downloader/handlers/http11.py]: We should not use assert statements in production code, outside tests, where we know the assertion may fail. Assert statements [are removed when optimizations are enabled](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement).

If you believe the assert can raise an `AssertionError` under some circumstances, better use an `if` statement instead.

File: docs/topics/request-response.rst [hunk 3]
```diff
@@ -611,6 +611,9 @@ Response objects
         This represents the :class:`Request` that generated this response.
     :type request: :class:`Request` object
 
+    :param certificate: an object representing the server's SSL certificate
+    :type certificate: :class:`twisted.internet.ssl.Certificate` object
```
Comment (Gallaecio) [inline docs/topics/request-response.rst]:
```suggestion
    :type certificate: twisted.internet.ssl.Certificate
```

I believe Sphinx already takes care of the linking

File: docs/topics/request-response.rst [hunk 4]
```diff
@@ -679,6 +682,11 @@ Response objects
         they're shown on the string representation of the Response (`__str__`
         method) which is used by the engine for logging.
 
+    .. attribute:: Response.certificate
+
+        A :class:`twisted.internet.ssl.Certificate` object representing
+        the server's SSL certificate
```
Comment (Gallaecio) [inline docs/topics/request-response.rst]:
```suggestion
        the server's SSL certificate.
```

File: scrapy/core/downloader/handlers/http11.py (hunk 2 above)
Comment (elacuesta) [reply scrapy/core/downloader/handlers/http11.py]: Updated, thanks :raised_hands:

File: scrapy/core/downloader/handlers/http11.py [hunk 5]
```diff
@@ -445,6 +446,12 @@ def __init__(self, finished, txresponse, request, maxsize, warnsize, fail_on_dat
         self._fail_on_dataloss_warned = False
         self._reached_warnsize = False
         self._bytes_received = 0
+        self._certificate = None
+
+    def connectionMade(self):
+        if self._certificate is None:
+            with suppress(AttributeError):
```
Comment (Gallaecio) [inline scrapy/core/downloader/handlers/http11.py]: TIL, nice!

File: docs/conf.py [hunk 6]
```diff
@@ -286,6 +286,7 @@
     'sphinx': ('https://www.sphinx-doc.org/en/master', None),
     'tox': ('https://tox.readthedocs.io/en/latest', None),
     'twisted': ('https://twistedmatrix.com/documents/current', None),
+    'twistedapi': ('https://twistedmatrix.com/documents/current/api', None),
```
Comment (kmike) [inline docs/conf.py]: I think it'd be better to reiview & merge it separately (https://github.com/scrapy/scrapy/pull/4197)

Comment (elacuesta) [reply docs/conf.py]: Good point, I added it here because otherwise the `twisted.internet.ssl.Certificate` references do not create links, but there is no harm in that for now.
